File: packages/git-core/src/plan/refs.ts
/**
 * Planners for repository metadata that is not a status or a diff.
 *
 * These reads all share one property: they answer "what does this repository
 * *say* about itself" without touching the working tree — configured remotes,
 * `.gitmodules`, and the tree entry a commit recorded for a path.
 *
 * `.gitmodules` is read with `git config --file` rather than by parsing the file:
 * Git's config grammar allows includes, escapes and repeated keys, and a
 * hand-written parser would disagree with Git on exactly the repositories that use
 * those features. The read is scoped to one file inside the worktree and never
 * loads the system or global config.
 */
import type { GitCommandSpec } from "../ports.js";
import type { PlanContext } from "./status.js";

function spec(
  context: PlanContext,
  argv: readonly string[],
  description: string,
  deadlineClass: GitCommandSpec["deadlineClass"] = "readonly",
): GitCommandSpec {
  return { argv, cwdHandle: context.cwdHandle, deadlineClass, description };
}

/**
 * Configured remotes and their URLs.
 *
 * `-v` prints `<name>\t<url> (fetch|push)` per remote. URLs are read, never
 * echoed: the host redacts credentials before a URL reaches a response, and a
 * remote's URL is display data, never an argument the browser can supply.
 */
export function planRemotes(context: PlanContext): GitCommandSpec {
  return spec(context, ["remote", "-v"], "remote -v");
}

/**
 * Submodule names, paths and URLs, as written in the worktree's `.gitmodules`.
 *
 * `-z` frames `key\nvalue` records with NUL, so a URL or path containing a newline
 * cannot be split into two entries. The caller checks that the file exists first:
 * Git reports a missing config file as an error, and "no `.gitmodules`" is an
 * ordinary answer, not a failure.
 */
export function planSubmoduleConfig(context: PlanContext): GitCommandSpec {
  return spec(
    context,
    ["config", "-z", "--file", ".gitmodules", "--get-regexp", "^submodule\\."],
    "config --file .gitmodules --get-regexp",
  );
}

/**
 * The tree entry a commit recorded for one path.
 *
 * This is the parent repository's *recorded* object for a submodule, which is the
 * first of the three object names a submodule row must show. The path is
 * literal-pathspec'd and placed after `--`, and it arrives as text only because
 * the host has already refused any path whose bytes cannot be represented.
 */
export function planLsTreeEntry(
  context: PlanContext,
  options: { readonly revision: string; readonly path: string },
): GitCommandSpec {
  return spec(
    context,
    [
      "--literal-pathspecs",
      "ls-tree",
      "-z",
      options.revision,
      "--",
      options.path,
    ],
    "ls-tree for one path",
  );
}

/**
 * Verify that a name really resolves to an object of the expected kind.
 *
 * Snapshot tips are object names captured earlier; between capture and use a ref
 * may move or an object may be pruned. `--verify --quiet` makes that a non-zero
 * exit instead of a confusing parse error later, and `^{commit}` refuses a tag or
 * a tree that merely happens to be named like a commit.
 */
export function planRevParseCommit(
  context: PlanContext,
  objectName: string,
): GitCommandSpec {
  return spec(
    context,
    ["rev-parse", "--verify", "--quiet", `${objectName}^{commit}`],
    "rev-parse --verify <name>^{commit}",
  );
}

/**
 * Check whether one object name is present locally.
 *
 * Used for the "missing parent" question: a graph that draws a parent as a root
 * hides a shallow clone or a partially fetched repository. `cat-file -e` answers
 * presence without transferring the body.
 */
export function planCatFileExists(
  context: PlanContext,
  objectNames: readonly string[],
): GitCommandSpec {
  if (objectNames.length === 0) {
    throw new Error("planCatFileExists requires at least one object name");
  }
  const stdin = new Uint8Array(
    [...`${objectNames.join("\n")}\n`].map((character) =>
      character.charCodeAt(0),
    ),
  );
  return {
    argv: ["cat-file", "--batch-check=%(objectname)"],
    cwdHandle: context.cwdHandle,
    deadlineClass: "readonly",
    description: "cat-file --batch-check",
    stdin,
  };
}

/** Hex-only object enumeration; revisions and user options are not accepted. */
export function planDisambiguateCommitPrefix(
  context: PlanContext,
  prefix: string,
): GitCommandSpec {
  if (!/^[0-9a-f]{4,64}$/.test(prefix))
    throw new Error("invalid object ID prefix");
  return spec(
    context,
    ["rev-parse", `--disambiguate=${prefix}`],
    "resolve commit prefix",
  );
}

/** Classify already enumerated objects without transferring their bodies. */
export function planCatFileObjectTypes(
  context: PlanContext,
  oids: readonly string[],
): GitCommandSpec {
  if (oids.length === 0)
    throw new Error("object type check requires candidates");
  for (const oid of oids) requireFullOid(oid);
  return {
    ...spec(
      context,
      ["cat-file", "--batch-check=%(objectname) %(objecttype)"],
      "check candidate object types",
    ),
    stdin: new Uint8Array(
      [...`${oids.join("\n")}\n`].map((character) => character.charCodeAt(0)),
    ),
  };
}

/** Ancestry between two resolved full object IDs. */
export function planIsCommitAncestor(
  context: PlanContext,
  ancestorOid: string,
  descendantOid: string,
): GitCommandSpec {
  requireFullOid(ancestorOid);
  requireFullOid(descendantOid);
  return spec(
    context,
    ["merge-base", "--is-ancestor", ancestorOid, descendantOid],
    "check commit ancestry",
  );
}

function requireFullOid(oid: string): void {
  if (!/^(?:[0-9a-f]{40}|[0-9a-f]{64})$/.test(oid))
    throw new Error("expected a full object ID");
}

File: crates/refyard-core/src/plan/refs.rs
//! Ref and layout planners that were not part of the status module.
//!
//! Kept separate so `plan/status.rs` stays about one panel: what a branch is called
//! and where a repository's directories are is refs work, even though both read the
//! same repository.

use super::{names_stdin, GitPlan};
use crate::bytes::is_object_name;
use crate::problem::CoreError;

/// The `.gitmodules` config entries, with NUL-framed key/value records.
pub fn plan_submodule_config() -> GitPlan {
    GitPlan::read(vec![
        "config".to_string(),
        "-z".to_string(),
        "--file".to_string(),
        ".gitmodules".to_string(),
        "--get-regexp".to_string(),
        "^submodule\\.".to_string(),
    ])
}

/// The recorded tree entries for an explicit set of literal repository paths.
///
/// Paths are arguments only after `--literal-pathspecs` and `--`; an invalid path or
/// non-object revision is refused here rather than becoming Git's option syntax.
pub fn plan_ls_tree_entries(revision: &str, paths: &[String]) -> Result<GitPlan, CoreError> {
    if !is_object_name(revision.as_bytes()) {
        return Err(CoreError::invalid_input(
            "plan_ls_tree_entries requires a 40- or 64-character lowercase object name",
        ));
    }
    if paths.is_empty() {
        return Err(CoreError::invalid_input(
            "plan_ls_tree_entries requires at least one path",
        ));
    }
    for path in paths {
        if path.is_empty()
            || path.starts_with('/')
            || path.contains('\0')
            || path
                .split('/')
                .any(|segment| segment.is_empty() || segment == "." || segment == "..")
        {
            return Err(CoreError::invalid_input(
                "plan_ls_tree_entries requires normalized repository-relative paths",
            ));
        }
    }
    let mut argv = vec![
        "--literal-pathspecs".to_string(),
        "ls-tree".to_string(),
        "-z".to_string(),
        revision.to_string(),
        "--".to_string(),
    ];
    argv.extend(paths.iter().cloned());
    Ok(GitPlan::read(argv))
}

/// Resolves a name that should be a commit, refusing to treat a tag object as one.
pub fn plan_rev_parse_commit(revision: &str) -> GitPlan {
    GitPlan::read(vec![
        "rev-parse".to_string(),
        "--verify".to_string(),
        "--quiet".to_string(),
        format!("{revision}^{{commit}}"),
    ])
}

/// Whether one commit is an ancestor of another. Exit 1 means "no", which is an
/// answer rather than a failure.
pub fn plan_is_commit_ancestor(ancestor: &str, descendant: &str) -> GitPlan {
    GitPlan::read(vec![
        "merge-base".to_string(),
        "--is-ancestor".to_string(),
        ancestor.to_string(),
        descendant.to_string(),
    ])
}

/// The configured remotes, with fetch and push URLs.
pub fn plan_remotes() -> GitPlan {
    GitPlan::read(vec!["remote".to_string(), "-v".to_string()])
}

/// Presence of each object name, answered without transferring a body.
///
/// Used for the "missing parent" question: a graph that draws a parent as a root hides
/// a shallow clone or a partially fetched repository, and `--batch-check` answers
/// presence for a whole list in one process.
pub fn plan_cat_file_exists(object_names: &[&str]) -> Result<GitPlan, CoreError> {
    if object_names.is_empty() {
        return Err(CoreError::invalid_input(
            "plan_cat_file_exists requires at least one object name",
        ));
    }
    let mut plan = GitPlan::new(vec![
        "cat-file".to_string(),
        "--batch-check=%(objectname)".to_string(),
    ]);
    plan.stdin = names_stdin(object_names);
    Ok(plan)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::plan::DeadlineClass;

    #[test]
    fn a_commit_lookup_requires_a_commit() {
        // Without the `^{commit}` peel, a tag object would resolve and then be read
        // as if it were a commit.
        let plan = plan_rev_parse_commit("refs/tags/v1");
        assert_eq!(
            plan.argv.last().map(String::as_str),
            Some("refs/tags/v1^{commit}")
        );
        assert_eq!(plan.deadline_class, DeadlineClass::Read);
    }

    #[test]
    fn ancestry_is_asked_as_a_question_not_a_failure() {
        let plan = plan_is_commit_ancestor("a", "b");
        assert_eq!(plan.argv, vec!["merge-base", "--is-ancestor", "a", "b"]);
    }

    #[test]
    fn presence_is_asked_for_a_whole_list_in_one_process() {
        // One `cat-file` for the page's parents, not one process per parent: a shallow
        // clone's page can name dozens of them.
        let plan = plan_cat_file_exists(&["aaaa", "bbbb"]).expect("plan");
        assert_eq!(plan.argv, vec!["cat-file", "--batch-check=%(objectname)"]);
        assert_eq!(plan.stdin, b"aaaa\nbbbb\n".to_vec());
        assert_eq!(plan.deadline_class, DeadlineClass::Read);
    }

    #[test]
    fn a_presence_check_with_no_names_is_refused() {
        // `cat-file --batch-check` with no input would wait for stdin that never comes.
        let error = plan_cat_file_exists(&[]).expect_err("empty");
        assert!(matches!(error, CoreError::InvalidInput { .. }));
    }

    #[test]
    fn submodule_config_is_scoped_to_gitmodules_and_nul_delimited() {
        let plan = plan_submodule_config();
        assert_eq!(
            plan.argv,
            vec![
                "config",
                "-z",
                "--file",
                ".gitmodules",
                "--get-regexp",
                "^submodule\\.",
            ]
        );
        assert_eq!(plan.deadline_class, DeadlineClass::Read);
        assert!(plan.stdin.is_empty());
    }

    #[test]
    fn recorded_submodule_entries_are_literal_and_cannot_escape_the_tree() {
        let oid = "0123456789abcdef0123456789abcdef01234567";
        let plan = plan_ls_tree_entries(oid, &["modules/child".into(), "-name[*]".into()])
            .expect("literal pathspecs");
        assert_eq!(
            plan.argv,
            vec![
                "--literal-pathspecs",
                "ls-tree",
                "-z",
                oid,
                "--",
                "modules/child",
                "-name[*]",
            ]
        );
        assert_eq!(plan.deadline_class, DeadlineClass::Read);
        assert!(plan.stdin.is_empty());
        assert!(plan_ls_tree_entries(oid, &["../outside".into()]).is_err());
        assert!(plan_ls_tree_entries("HEAD", &["modules/child".into()]).is_err());
    }
}
